x/ibc/23-commitment/store.go

```diff
@@ -0,0 +1,99 @@
+package commitment
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+// Store proves key-value pairs' inclusion or non-inclusion with
+// the stored commitment proofs against the commitment root.
+type Store interface {
+	Prove(key, value []byte) bool
+}
+
+var _ Store = prefix{} // TODO: pointer
+
+type prefix struct {
+	store  Store
+	prefix []byte
+}
+
+func NewPrefix(store Store, pref []byte) prefix {
+	return prefix{
+		store:  store,
+		prefix: pref,
+	}
+}
+
+func (prefix prefix) Prove(key, value []byte) bool {
+	return prefix.store.Prove(join(prefix.prefix, key), value)
+}
+
+var _ Store = (*store)(nil)
+
+type store struct {
+	root     Root
+	path     Path
+	proofs   map[string]Proof
+	verified map[string][]byte
+}
+
+// NewStore constructs a new Store with the root, path, and proofs.
+// The proofs are not proven immediately because proofs require value bytes to verify.
+// If the kinds of the arguments don't match, returns error.
+func NewStore(root Root, path Path, proofs []Proof) (res *store, err error) {
```

Comment:
Hmm, when do we end up constructing a store with a list of proofs like this?

Reply:
Is this designed to serve as a cache of verified proofs?

Reply:
The `store` works as a key-proof store which has only one method `Verify`. For all `verifyMembership(root, proof, path, value)` in the spec is translated into `store.Prove(path, value)`, and `verifyNonMembership(root, proof, path)` into `store.Prove(path, nil)` where `store = NewStore(root, _, []Proof{proof})`. This gives kvstore like interface that the higher abstraction can use easily.